### PR TITLE
Evaluation branch

### PR DESCRIPTION
**Pull request by bharat-suri**

### CODE REVIEW

File: gsoc2018-bharat/src/sparql.py [hunk 1]
```diff
@@ -2,20 +2,25 @@
...
+
 query = """
-    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+    PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>
+    PREFIX rdf:<http://www.w3.org/1999/02/22-rdf-syntax-ns#>
     PREFIX dbo: <http://dbpedia.org/ontology/>
-    SELECT DISTINCT ?s ?abstract WHERE {
+    SELECT DISTINCT ?s ?abstract ?url ?p
+    WHERE {
         ?s a/rdfs:subClassOf* dbo:Organisation .
```
Comment (mommi84) [inline gsoc2018-bharat/src/sparql.py]: As discussed via call, I suggest to split the query into two queries and merge their results later. Having the abstracts multiplied by the number of classes isn't a good idea.

File: gsoc2018-bharat/src/create_db.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import Encoder
+import logging
+
+
+logging.basicConfig(format='%(levelname)s : %(message)s', level=logging.INFO)
+logging.root.level = logging.INFO
+model = torch.load('model/description_encoder')
+m = FastText.load('model/entity_fasttext_n100')
+wv = m.wv
+del m
```
Comment (mommi84) [inline gsoc2018-bharat/src/create_db.py:16]: Is the creation of the `m` object necessary?

Comment (bharat-suri) [reply gsoc2018-bharat/src/create_db.py:16]: Yes. I need to load all the pre-trained embeddings into the main memory. I delete it because I only need the keyed vectors, not the entire model. I can make use of GloVe instead of FastText to save the hassle of loading the model first.

File: gsoc2018-bharat/README.md [hunk 3]
```diff
@@ -2,16 +2,27 @@
 
 This project aims to generate complex word embeddings for Out-Of-Vocabulary entities. After completion, the package will be able to generate pre-trained embeddings, improve them by generating embeddings on-the-fly, evaluate the benchmarks using pre-trained embeddings, and make the same evaluations on the imporoved embeddings.
 
-## Requirements
+### Getting started
 
-All the libraries used under this project are included in the file, requirements.txt. To install, just run the command
```
Comment (tgalery) [inline gsoc2018-bharat/README.md:7]: I have the feeling that these changes belong to a previous version of the repo, so maybe we have to undo them

File: gsoc2018-bharat/README.md [hunk 4]
```diff
@@ -2,16 +2,27 @@
...
+Here is how you can start training and testing the model yourself. First, clone the repository on your local machine. I assume you are familiar with Git, and have it installed on your system prior to using this repo.
 
 ```shell
+$ git clone git@github.com:tramplingWillow/embeddings.git
```
Comment (tgalery) [inline gsoc2018-bharat/README.md]: for example, this would be `dbpedia/embeddings.git` since we are merging to dbpedia, right ?

File: gsoc2018-bharat/src/create_db.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import Encoder
+import logging
+import sys
+
+
+logging.basicConfig(format='%(levelname)s : %(message)s', level=logging.INFO)
+logging.root.level = logging.INFO
+model = torch.load('model/description_encoder')
+# wv = KeyedVectors.load_word2vec_format('model/glove.6B.100d.txt.word2vec',
+#                                        binary=False)
```
Comment (tgalery) [inline gsoc2018-bharat/src/create_db.py]: remove the commented code

File: gsoc2018-bharat/src/create_db.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+m = FastText.load('model/entity_fasttext_n100')
+wv = m.wv
+del m
+punctuation = '!"#$%&\'()*+,.:;<=>?@[\\]^`{|}~'
+table = str.maketrans('', '', punctuation)
+
+
+def main(model, wv, abstracts_file, db_file):
```
Comment (tgalery) [inline gsoc2018-bharat/src/create_db.py:21]: It feels like a bunch of this code lives somewhere else. Have I seen this before. Also you need to decide whether you are using print statements to debug or not

File: gsoc2018-bharat/src/create_db.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
+import json
+from gensim.models import FastText
+from gensim.models import KeyedVectors
+import torch
+from torch.autograd import Variable
+import numpy as np
+import Encoder
```
Comment (tgalery) [inline gsoc2018-bharat/src/create_db.py]: I think `Encoder` has been removed / renamed, so maybe you need to reflect the change here

File: gsoc2018-bharat/src/evaluate.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+logging.basicConfig(format='%(levelname)s : %(message)s', level=logging.INFO)
+logging.root.level = logging.INFO
+punctuation = '!"#$%&\'()*+,.:;<=>?@[\\]^`{|}~'
+table = str.maketrans('', '', punctuation)
+dictionary = {}
+model = FastText.load('model/entity_fasttext_n100')
+wv = model.wv
+del model
+# f = '/Volumes/Seagate/SeagateBackupPlus/Github/GSoC2018/glove.6B/glove.6B.100d.txt'
```
Comment (tgalery) [inline gsoc2018-bharat/src/evaluate.py]: Ok, so you need to thoroughly review this before you PR. General issues:

1. Lots of commented out code 
2. Lots of `global` uses of variables within a function when the function docstring mentions that the variable is passed as an arg . You need to decide which is the case, passing data as args is usually much better than declaring global states

File: gsoc2018-bharat/src/evaluate.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        except KeyError:
+            itr -= 1
+            continue
+        if np.any(np.isnan(r)):
+            itr -= 1
+            continue
+        else:
+            if r.ndim == 2:
+                try:
+                    r = r.mean(axis=0).reshape(1, -1)
```
Comment (tgalery) [inline gsoc2018-bharat/src/evaluate.py]: This whole function as it stands is very hard to understand. Probably you can isolate this whole code block into a function and document what each index means. Also use way more descriptive names for your varirables, e.g. `random_vec` instead of `vec5`, `zero_vec` instead of `vec6`

File: gsoc2018-bharat/src/evaluate.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+                except (ValueError, KeyError) as _:
+                    itr -= 1
+                    continue
+            else:
+                itr -= 1
+                continue
+    scores = scores / norm(scores)
+    cos_scores = cos_scores / norm(cos_scores)
+    print(scores)
+    print(cos_scores)
```
Comment (tgalery) [inline gsoc2018-bharat/src/evaluate.py]: maybe the some data strucutre should be returned, and then you can code up a `print_summary` function that displays that information somehow.

File: gsoc2018-bharat/src/sparql.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
+from SPARQLWrapper import SPARQLWrapper, JSON
+import asyncio
```
Comment (tgalery) [inline gsoc2018-bharat/src/sparql.py:2]: also have the feeling that i saw this somewhere else. Is this currently used to get training / eval data ?

File: gsoc2018-bharat/src/tsne.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
+import numpy as np
+from sklearn.manifold import TSNE
```
Comment (tgalery) [inline gsoc2018-bharat/src/tsne.py:2]: make this a cli tool that takes the file you want to display instead of hardcoding the path

**Review by tgalery — CHANGES_REQUESTED** [review]

File: gsoc2018-bharat/README.md (hunk 4 above)
Comment (bharat-suri) [reply gsoc2018-bharat/README.md]: Yes.

File: gsoc2018-bharat/src/sparql.py (hunk 11 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/sparql.py:2]: It was used to get the eval data for t-SNE. I used the data to plot entities in the graphs that I shared on the slack channel.
